### PR TITLE
Add floating point mutators

### PR DESCRIPTION
**Pull request by oetr**

### CODE REVIEW

File: src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+        @Override
+        public Float mutate(Float value, PseudoRandom prng) {
+          return (float) mutateImpl(value, prng);
+        }
+
+        @Override
+        public Float read(DataInputStream in) throws IOException {
+          float value = (float) in.readFloat();
+          return value;
```
Comment (fmeum) [inline src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java:64]:
```suggestion
          return in.readFloat();
```

File: src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            @Override
+            public Double mutate(Double value, PseudoRandom prng) {
+              value = (double) mutateImpl(value, prng);
+              return value;
+            }
+
+            @Override
+            public Double read(DataInputStream in) throws IOException {
+              double value = (double) in.readDouble();
+              return value;
```
Comment (fmeum) [inline src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java:94]:
```suggestion
              return in.readDouble();
```

File: src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    @ForOverride protected abstract Float mutateWithLibFuzzer(Float value);
+
+    @Override
+    public T detach(T value) {
+      // Always immutable.
+      return value;
+    }
+
+    @Override
+    public String toDebugString(Predicate<Debuggable> isInCycle) {
```
Comment (fmeum) [inline src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java:129]: If you override this method, you might as well just return `"Double"`. Also applies to the float mutator.

File: src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+              out.writeDouble(value);
+            }
+          });
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  // Visible for testing.
+  static abstract class AbstractFloatMutator<T extends Float> extends SerializingMutator<T> {
```
Comment (fmeum) [inline src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java:108]: Not entirely sure why this class is abstract: If you have one class per type anyway, why not make it not abstract and merge in the implementation of `SerializingMutator` defined anonymously above?

File: src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java (hunk 3 above)
Comment (oetr) [reply src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java:129]: Done!

File: src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java (hunk 4 above)
Comment (oetr) [reply src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java:108]: I initially thought that the two types can be combined in one. I merged the methods, though still have trouble with types.

File: src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java (hunk 1 above)
Comment (oetr) [reply src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java]: Changed, but see the new version---I feel like the number casts can be reduced, but not sure how.

File: src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java (hunk 2 above)
Comment (oetr) [reply src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java]: Same as with floats.

File: src/main/java/com/code_intelligence/jazzer/mutation/engine/SeededPseudoRandom.java [hunk 5]
```diff
@@ -118,6 +118,18 @@ public long closedRange(long lowerInclusive, long upperInclusive) {
     }
   }
 
+  public float closedRange(float lowerInclusive, float upperInclusive) {
+    require(lowerInclusive <= upperInclusive);
+    return (float) random.nextDouble(-Float.MAX_VALUE, Float.MAX_VALUE);
```
Comment (fmeum) [inline src/main/java/com/code_intelligence/jazzer/mutation/engine/SeededPseudoRandom.java:123]: This is rather confusing: The function ignores its arguments. You could either get rid of the arguments or honor them.

File: src/main/java/com/code_intelligence/jazzer/mutation/engine/SeededPseudoRandom.java [hunk 6]
```diff
@@ -118,6 +118,18 @@ public long closedRange(long lowerInclusive, long upperInclusive) {
...
+  }
+
+  public double closedRange(double lowerInclusive, double upperInclusive) {
+    require(lowerInclusive <= upperInclusive);
+    // using "random.nextDouble(-Double.MAX_VALUE, Double.MAX_VALUE);" is not possible. TODO: find
+    // out why
+    return random.nextDouble(-1e100, 1e100);
```
Comment (fmeum) [inline src/main/java/com/code_intelligence/jazzer/mutation/engine/SeededPseudoRandom.java:130]: See above.

File: src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    @Override
+    public Float read(DataInputStream in) throws IOException {
+      return in.readFloat();
+    }
+
+    @Override
+    public void write(Float value, DataOutputStream out) throws IOException {
+      out.writeFloat((float) value);
```
Comment (fmeum) [inline src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java:75]: This should also compile:
```suggestion
      out.writeFloat(value);
```

File: src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    @Override
+    public Double read(DataInputStream in) throws IOException {
+      return in.readDouble();
+    }
+
+    @Override
+    public void write(Double value, DataOutputStream out) throws IOException {
+      out.writeDouble((double) value);
```
Comment (fmeum) [inline src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java:114]: This should also compile:
```suggestion
      out.writeDouble(value);
```

Comment (oetr) [reply src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java]: Done!

File: src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java (hunk 7 above)
Comment (oetr) [reply src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java]: Done!

File: src/main/java/com/code_intelligence/jazzer/mutation/engine/SeededPseudoRandom.java (hunk 6 above)
Comment (oetr) [reply src/main/java/com/code_intelligence/jazzer/mutation/engine/SeededPseudoRandom.java]: Moved the code to the mutator.

File: src/main/java/com/code_intelligence/jazzer/mutation/engine/SeededPseudoRandom.java (hunk 5 above)
Comment (oetr) [reply src/main/java/com/code_intelligence/jazzer/mutation/engine/SeededPseudoRandom.java]: Moved the code to the mutator.

**Review by fmeum — APPROVED** [review]

**Pull request merged**